modules/programs/htop.nix

```diff
@@ -0,0 +1,331 @@
+    TTY_NR = 6;
+    TPGID = 7;
+    MINFLT = 9;
+    MAJFLT = 11;
+    PRIORITY = 17;
+    NICE = 18;
+    STARTTIME = 20;
+    PROCESSOR = 37;
+    M_SIZE = 38;
+    M_RESIDENT = 39;
+    ST_UID = 45;
+    PERCENT_CPU = 46;
+    PERCENT_MEM = 47;
+    USER = 48;
+    TIME = 49;
+    NLWP = 50;
+    TGID = 51;
+    CMINFLT = 10;
+    CMAJFLT = 12;
+    UTIME = 13;
+    STIME = 14;
+    CUTIME = 15;
+    CSTIME = 16;
+    M_SHARE = 40;
+    M_TRS = 41;
+    M_DRS = 42;
+    M_LRS = 43;
+    M_DT = 44;
+    CTID = 99;
+    VPID = 100;
+    VXID = 102;
+    RCHAR = 102;
+    WCHAR = 103;
+    SYSCR = 104;
+    SYSCW = 105;
+    RBYTES = 106;
+    WBYTES = 107;
+    CNCLWB = 108;
+    IO_READ_RATE = 109;
+    IO_WRITE_RATE = 110;
+    IO_RATE = 111;
+    CGROUP = 112;
+    OOM = 113;
+    IO_PRIORITY = 114;
+  };
+
+  # Mapping from names to defaults
+  meters = {
+    Clock = 2;
+    LoadAverage = 2;
+    Load = 2;
+    Memory = 1;
+    Swap = 1;
+    Tasks = 2;
+    Uptime = 2;
+    Battery = 2;
+    Hostname = 2;
+    AllCPUs = 1;
+    AllCPUs2 = 1;
+    LeftCPUs = 1;
+    RightCPUs = 1;
+    LeftCPUs2 = 1;
+    RightCPUs2 = 1;
+    Blank = 2;
+    CPU = 1;
+    "CPU(1)"= 1;
+    "CPU(2)" = 1;
+    "CPU(3)" = 1;
+    "CPU(4)" = 1;
+  };
+
+  singleMeterType = types.either (types.enum (attrNames meters)) (types.submodule {
+    options = {
+      kind = mkOption {
+        type = types.enum (attrNames meters);
+        example = "AllCPUs";
+        description = "What kind of meter";
+      };
+
+      mode = mkOption {
+        type = types.enum [ 1 2 3 4 ];
+        example = 2;
+        description = "Which mode the meter should use, one of 1(Bar) 2(Text) 3(Graph) 4(LED)";
+      };
+    };
+  });
+
+  # Converts the meter to an attribute set if it isn't already
+  meterStandard = m: if builtins.isString m then {
+    kind = m;
+    mode = meters.${m};
+  } else m;
+
+
+  meterType = types.submodule {
+    options = {
+      left = mkOption {
+        description = "Left meters";
+        default = [ "AllCPUs" "Memory" "Swap" ];
+        example = [
+          "Memory"
+          "LeftCPUs2"
+          "RightCPUs2"
+          { kind = "CPU"; mode = 3; }
+        ];
+        type = types.listOf singleMeterType;
+      };
+      right = mkOption {
+        description = "Right meters";
+        default = [ "Tasks" "LoadAverage" "Uptime" ];
+        example = [
+          { kind = "Clock"; mode = 4; }
+          "Uptime"
+          "Tasks"
+        ];
+        type = types.listOf singleMeterType;
+      };
+    };
+  };
+in
+{
+  options.programs.htop = {
+    enable = mkEnableOption "Enable htop";
+
+    fields = mkOption {
+      type = types.listOf (types.enum (attrNames fields));
+      default = [ "PID" "USER" "PRIORITY" "NICE" "M_SIZE" "M_RESIDENT" "M_SHARE" "STATE" "PERCENT_CPU" "PERCENT_MEM" "TIME" "COMM" ];
+      example = [ "PID" "USER" "PRIORITY" "PERCENT_CPU" "M_RESIDENT" "PERCENT_MEM" "TIME" "COMM" ];
+      description = "Fields";
```

Comment:
Maybe expand description slightly. Something like "Active fields of process data." or whatever.